mllib/src/main/scala/org/apache/spark/ml/feature/VectorIndexer.scala

```diff
@@ -311,22 +342,39 @@ class VectorIndexerModel private[ml] (
   // TODO: Check more carefully about whether this whole class will be included in a closure.
 
   /** Per-vector transform function */
-  private val transformFunc: Vector => Vector = {
+  private lazy val transformFunc: Vector => Vector = {
```

Comment:
We need `lazy` modifier here. Otherwise we cannot get correct `$(handleInvalid)` value in this `transformFunc`, because of var initialization order.

Reply:
Or we could change it to a def here in order to avoid methods-in-methods.

Reply:
I don't think changing it to a `def` is a good idea.
Because, it will cause the whole `Model` object to be serialized.
Current code, it will create a closure, which only involve the outer variables which are needed. It can help to avoid unexpected serialization.
```
lazy val transformFunc = {
    // the returned closure `f` will only reference the following 4 outer variables.
    val sortedCatFeatureIndices = categoryMaps.keys.toArray.sorted
    val localVectorMap = categoryMaps
    val localNumFeatures = numFeatures
    val localHandleInvalid = getHandleInvalid
    val f : Vector => Vector = { (v: Vector) => ...}
    f   // return closure `f`
}
```

Reply:
Can I ask about the lazy val here? What happens here if a user sets the params, calls transform, modifies one of the params, and then calls transform again? Is that a concern?

Reply:
I'm not super familiar with scala closures, but if we do
```
def transformFunc: Vector => Vector = {
    
    val sortedCatFeatureIndices = categoryMaps.keys.toArray.sorted
    val localVectorMap = categoryMaps
    val localNumFeatures = numFeatures
    val localHandleInvalid = getHandleInvalid
    val f : Vector => Vector = { (v: Vector) => ...}
    f
}
```
Won't we still get a closure and avoid serializing the entire model?

Reply:
@MrBago Use `lazy val` will generate the closure only once. If use `def`, each time call `transformFunc` will generate the closure again.
What you concern is when user setParams for this Model, then the param won't work. But, I think this is not a issue, because `Model` do not allow user to set params, the params are copied from estimator, isn't it ?

Reply:
@WeichenXu123 Thanks for the clarification, I'm still working on understanding Params and when they can & can't be modified :).